Lib/test/test_shutil.py

```diff
@@ -1259,6 +1259,30 @@ def test_copyfile_nonexistent_dir(self):
         write_file(src_file, 'foo')
         self.assertRaises(FileNotFoundError, shutil.copyfile, src_file, dst)
 
+    def test_copyfile_copy_dir(self):
+        # Issue 45234
+        # test copy() and copyfile() raising proper exceptions when src and/or
+        # dst are directories
+        src_dir = self.mkdtemp()
+        src_file = os.path.join(src_dir, 'foo')
+        dir2 = self.mkdtemp()
+        dst = os.path.join(src_dir, 'does_not_exist/')
+        write_file(src_file, 'foo')
+        if sys.platform == "win32":
+            err = PermissionError
+        else:
+            err = IsADirectoryError
+
+        self.assertRaises(err, shutil.copyfile, src_dir, dst)
+        self.assertRaises(err, shutil.copyfile, src_file, src_dir)
+        self.assertRaises(err, shutil.copyfile, dir2, src_dir)
+        self.assertRaises(err, shutil.copy, dir2, src_dir)
+        self.assertRaises(err, shutil.copy2, dir2, src_dir)
```

Comment:
Shouldn't this just test `copyfile()`? 

Note that for Windows there's been some discussion about changing `copy2()`, and maybe `copy()`, to call the system `CopyFileExW()` function, which also copies alternate data streams, extended attributes, and security resource attributes. In that case the high-level copy functions will no longer use `copyfile()`, `copymode()`, and `copystat()` in Windows.

Reply:
I think it's worth testing copy and copy2 for consistency of handling a dir arg, because they already do some dir-related logic (for dst arg), but this may be done here or in separate tests. Looking at other tests, to follow the same naming pattern, we can do `test_copy_dirs()` and `test_copy2_dirs()`.

But it makes sense to do this testing in the same function for all three because we're ensuring consistency and also it's convenient because the set up of dirs and expected errors is the same.

I can remove these lines though if you think it's not worth testing them here at this point in view of the other work you mentioned.

Reply:
Currently what `copy()` and `copy2()` do that's related to `src` being a directory is implemented in `copyfile()`, so only the latter really needs to be tested. I see the point about checking consistency, but would separate tests maybe be more obvious for someone who's modifying `copy()` or `copy2()`?

Reply:
Sounds good, I will extract them into separate tests a bit later today.

Reply:
I've updated the tests and they're passing, please take a look..